### ISSUE TITLE
Implement Fisher-weighted SVD compression

### ISSUE THREAD

**Issue by jiahao** [rai-llc/TransformerREPL.jl#8]
Paper references (Thanks @evanmiller)

- [Numerical Optimizations for Weighted Low-rank Estimation on Language Model](https://arxiv.org/abs/2211.09718)  - looks like these authors have rediscovered the alternating projections method of [(Higham, 2002, Algorithm 3.3)](https://eprints.maths.manchester.ac.uk/232/1/paper3.pdf)
- [Language model compression with weighted low-rank factorization](https://arxiv.org/abs/2207.00112)

An explicit formula for the Fisher weight matrix for the cross-entropy loss (which is [used in llama](https://github.com/huggingface/transformers/blob/892399c5ffdc3c7d0e4f9b2c1c81511293c1581b/src/transformers/models/llama/modeling_llama.py#L712)) is given in §A.2 of
- [Limitations of the Empirical Fisher Approximation for Natural Gradient Descent](https://arxiv.org/abs/1905.12558)

**Comment by evanmiller** [comment]
See also their patent application https://patents.google.com/patent/US20230106213A1/en

**Comment by jiahao** [comment]
It looks like the weighted Hadamard generalization of the SVD has already been solved in the [nearest correlation matrix problem](https://nhigham.com/2013/02/13/the-nearest-correlation-matrix/), for which [a Newton method is available](https://www.nag.com/numeric/cl/nagdoc_cl25/html/g02/g02ajc.html). That should be much more efficient than gradient descent or projected descent.

**Comment by jiahao** [comment]
Adding also the related [GPTQ](https://arxiv.org/abs/2210.17323) method for quantization, which uses the data Hessian, $XX'$, in effect computing the same empirical Fisher-based weight matrix, except using the least-squares loss instead. They use an ad hoc [modified Cholesky](https://nhigham.com/2020/12/22/what-is-a-modified-cholesky-factorization/) which can probably be improved upon with a more robust implementation. 

https://github.com/IST-DASLab/gptq
